### ISSUE TITLE
initializationOptions on project basis

### ISSUE THREAD

**Issue by Glaeqen** [kak-lsp/kak-lsp#480]
Hey!

I have a question regarding initalization options. Now, one can either use lsp\_server\_initalization\_options or use global kak\_lsp.conf directly. The problem I see is that lsp\_server\_initalization\_options is a `str-to-str-map` so I'm not sure how can I express value which is a JSON list. Like for example

```json
{
    "initializationOptions": {
        "cargo": {
            "features": ["feature1", "feature2"]
        }
    }
}
```

I managed to get it to work with a global kak\_lsp.conf via
```toml
[language.rust.initialization_options.cargo]
features = ["feature1", "feature2"]
```

but it is not that great really. Tbf, it would be nice to be able to have these to-be-enabled-initialization-options stored and read from somewhere so I can jump between projects and kak-lsp would pick these up automatically. Thanks for help! Amazing project btw <3

**Comment by krobelus** [comment]
Right, the `rust.clippy_preference="on"` example in our readme only mentions that maps and primitives are supported, but arrays aren't.

So currently we can dynamically set initialization options using

```
set-option global lsp_server_initialization_options cargo.foo=bar
```

`lsp-get-server-initialization-options` will convert that to valid TOML, which is loaded on-the-fly (and overrides kak-lsp.toml)

```toml
cargo.foo = "bar"
```

but for your use case we need to support something like

```
set-option global lsp_server_initialization_options cargo.features="feature1,feature2"
```

**Comment by Glaeqen** [comment]
Yep. I'm not sure what is a sane solution to this problem. 

Also, problem probably is even more general than this Kakoune option. It'd be nice to have a state-of-the-art approach (maybe even mentioned in wiki/readme) for initialization option management on a project basis. 

Can't help but notice that one of the first class citizens supported by kak-lsp is Rust. Especially in Rust Embedded projects, usage of feature flags is really intensive. In HAL projects these are very often used to indicate against which device project is being compiled. Having these reliably switchable on fly would be immensly helpful.

**Comment by krobelus** [comment]
Hold on, I think this already works. :joy_cat:  
I have this in my configuration:

```kak
set-option global lsp_server_configuration pyls.configurationSources=["pycodestyle","flake8"]
```

so as long as the thing right of the `=` is valid TOML you're good.

> It'd be nice to have a state-of-the-art approach (maybe even mentioned in wiki/readme) for initialization option management on a project basis.

Yeah sounds like something for the Wiki (the readme is already fairly dense).

> Having these reliably switchable on fly would be immensly helpful.

Yes, this already works. Whenenver `%opt{lsp_server_initialization_options}` changes, then `kak-lsp` will reconfigure the language server.

Really it's just a question of convention/preference how you implement it.
You can create a `.local.kak` in your project's root that contains something like

```kak
set-option global lsp_server_initialization_options cargo.features=["feature1","feature2"]
```

And then add `try %{ source .local.kak }` to your kakrc. (I didn't try this with cargo, but it should work)

Personally, I don't like scattering those files around my projects, so I just have a section in my main kakrc that does some project-specific initialization based on the current directory.
Here's how this looks like for some other options, but the same concept applies for lsp options.

```kak
define-command my-setup-for-kak-lsp -params 1 %{
	set-option global indentwidth 4
	hook -group my global WinSetOption filetype=rust %{
		set-option window autowrap_column 100
	}
	map global user m %{: write-all<ret>: cargo build<ret><esc>} -docstring 'cargo build'
}
define-command -override my-setup-for-fish-shell -params 1 %{
	set-option global makecmd "ninja -C %arg{1}/build fish"
	set-option global indentwidth 4
	hook -group my global WinSetOption filetype=(c|cpp) %{
		set-option window autowrap_column 100
	}
}
evaluate-commands %sh{
	root="$(git rev-parse --show-toplevel 2>/dev/null)"
	echo try %{ "my-setup-for-${root##*/}" "$root" } # call the setup function, if it exists
}
```

**Comment by Glaeqen** [comment]
It's interesting. I'm surprised a little by this trick with `.local.kak` file. I guess you have to lookup every parent directory in a chain from the file; sourcing itself does not seem to be enough IMO. Also, my understanding is that you're not using sessions in kakoune but rather open an editor everytime you want to edit something? When clients run in the same session, kakrc is read only on a session creation so it'd require using some hooks instead.

**Comment by Glaeqen** [comment]
Also, `set-option global lsp_server_initialization_options cargo.features=["feature1","feature2"]`, does not work. Init JSON produced by kak-lsp contains `"initializationOptions":{"cargo.features":["feature1","feature2"]}` which seemingly is ignored by rust-analyzer because it expects nested structure of `cargo` and `features`.

**Comment by iMarv** [comment]
Running into a similar issue with the Deno LSP.
In  deno I can use so called `Import maps` per project, which map packages to their corresponding url, which means I have one `import_map.json` file per project which needs to be passed to the lsp to correctly resolve import paths in my code.

In addition to that init option, I have to additional ones which enable the whole thing and allow lsp based linting.

Would really be nice to have an option on a per-project basis to allow passing the correct json file

**Comment by krobelus** [comment]
> Also, set-option global lsp_server_initialization_options cargo.features=["feature1","feature2"], does not work. Init JSON produced by kak-lsp contains "initializationOptions":{"cargo.features":["feature1","feature2"]} which seemingly is ignored by rust-analyzer because it expects nested structure of cargo and features.

Right, my bad.
Commit abe3244 (branch [`dynamic-config`](https://github.com/krobelus/kak-lsp/tree/dynamic-config)) has a tentative fix.
It allows to write something like

    set-option global lsp_server_initialization_options %sh{ cat project-init-options.toml }

(I'm not sure if we should break compat, or just add a new option.)

@Glaeqen @iMarv would that work for your use case?
Do you prefer to specify JSON instead of TOML? I think in the dynamic options we could support both by autodetecting the format.
JSON is uglier, but it might be convenient since it's used by others. Not sure if that's worth it in the long run.

>  guess you have to lookup every parent directory in a chain from the file;

or use something like `git rev-parse --show-toplevel`

> Also, my understanding is that you're not using sessions in kakoune

I'm using one session per project, so usually sourcing at startup is enough. (I'm also using an [idempotent config](https://discuss.kakoune.com/t/how-to-reload-config-without-leaving-kakoune/1586/3?u=krobelus), so I can source it again any time.)
So yeah, if you are really using a session for multiple projects with conflicting Cargo features you might want to use hooks.
(Alternatively, maybe just send the union of all features you might need. How often do they really conflict? Keep in mind that if you send unknown features, they are ignored. Anyway, this is just an idea, which I haven't tried.).

**Comment by iMarv** [comment]
I guess json or toml, or yaml or w/e does not really matter. I feel like toml is less of a pain to write as it requires far less indents and quotes.

So I am generally new to this, would your example above source a `project-init-options.toml` from the directory where kakoune is launched? If thats the case and if the lsp then resolves the paths relative to that file my problem would be solved

**Comment by krobelus** [comment]
> would your example above source a project-init-options.toml from the directory where kakoune is launched?

yep

> if the lsp then resolves the paths relative to that file

I'm not completely sure what you mean (any example?). `kak-lsp` doesn't know the path of `project-init-options.toml`, because we only pass the contents.
`kak-lsp` and the language server resolve most paths relative to the project root.
The project root is determined by going up the directory hierarchy, until one of the `roots` parameters (from `kak-lsp.toml`) matches. So if you use `roots = [".git"]`, it will use the Git root dir as project root. If none matches, `kak-lsp` falls back to the file's directory.

**Comment by krobelus** [comment]
I finally found time to look at this in detail. Turns out this was a simple bug, see 0e6fbd2.

TL;DR is that this will work now:
`set-option global lsp_server_initialization_options cargo.features=["feature1","feature2"]`

So I didn't add a direct option to load a TOML file, but it shouldn't be hard to do.
You can use the full power of Kakoune scripting to set `%opt{lsp_server_initialization_options}`.
Entries are of the form `key=value` where key is the dot-separated path and value must be a valid TOML value.
